src/ripple/app/main/Application.cpp

```diff
@@ -262,35 +258,42 @@ class ApplicationImp : public Application, public RootStoppable, public BasicApp
         std::unique_ptr<Config> config,
         std::unique_ptr<Logs> logs,
         std::unique_ptr<TimeKeeper> timeKeeper)
-        : RootStoppable("Application")
-        , BasicApp(numberOfThreads(*config))
+        : BasicApp(numberOfThreads(*config))
         , config_(std::move(config))
         , logs_(std::move(logs))
         , timeKeeper_(std::move(timeKeeper))
         , m_journal(logs_->journal("Application"))
 
         // PerfLog must be started before any other threads are launched.
-        , perfLog_(perf::make_PerfLog(
+        , perfLog_(perf::make_PerfLogImp(
               perf::setup_PerfLog(
                   config_->section("perf"),
                   config_->CONFIG_DIR),
-              *this,
               logs_->journal("PerfLog"),
-              [this]() { signalStop(); }))
+              [&] { signalStop(); }))
 
         , m_txMaster(*this)
 #ifdef RIPPLED_REPORTING
         , pgPool_(
               config_->reporting() ? make_PgPool(
                                          config_->section("ledger_tx_tables"),
-                                         *this,
                                          logs_->journal("PgPool"))
                                    : nullptr)
 #endif
 
-        , m_nodeStoreScheduler(*this)
+        , m_collectorManager(CollectorManager::New(
+              config_->section(SECTION_INSIGHT),
+              logs_->journal("Collector")))
+
+        , m_jobQueue(std::make_unique<JobQueue>(
+              m_collectorManager->group("jobq"),
+              logs_->journal("JobQueue"),
+              *logs_,
+              *perfLog_))
+
+        , m_nodeStoreScheduler(*m_jobQueue)
```

Comment:
The biggest changes to `NodeStoreScheduler` were made in [this commit](https://github.com/ripple/rippled/pull/3741/commits/50a99674bd0d96aba6553832033c4feca5f29d29), which is not very large. It makes it easy to see that `NodeStoreScheduler` never needed to be a `Stoppable`. It used `JobQueue`, but was not responsible for stopping `JobQueue`, and it had no resources (e.g. `std::thread`s) that needed to be stopped, thus it was not responsible for stopping anything, thus it did not need to be a `Stoppable`, even in the old framework where `Stoppable` existed. It was one of the first and easiest classes I could remove from the `Stoppable` hierarchy.

If you read my description for this PR, you'll see I'm trying to move to RAII, where starting and stopping is not just "like" pushing and popping from a stack, but enforced by the language. We agree on that. Right now, `Stoppable` is a mess. In `develop` right now, `JobQueue` starts in its constructor, but is stopped before anything else, and "stopped" again (a no-op) in its destructor (this second stop is to cover the case when the `ApplicationImp` is never stopped and thus doesn't stop the `JobQueue`). I'm trying to get it to be less of a mess, but I stopped short of perfect. I stopped at this point to build support for this direction. I want to reach RAII eventually, but getting there will require even more radical changes that will be easier to swallow if this changeset is the starting point, instead of the current state of `develop`.

Reply:
You seemed to have picked up on another mislead of `Stoppable`: that children depend on their parents. `NodeStoreScheduler` and `JobQueue` were a strange contradiction of that that I cannot explain. This changeset removes that contradiction.

Reply:
Got it, thanks for the explanation. I was mistaken about the stopping order of the various `Stoppable`s. I do think RAII would be best, and that what you have here is clearer than what's currently in `develop`.